doc: add documentation for adapters structure and calculated documents (#108)
Benchling-Open-Source/allotropy: src/allotropy/parsers/README.md

> # Allotropy parser structure

Reviewer: ```suggestion
# allotropy parser structure
```

> Allotropy is composed by multiple data parsers, each of them should follow a general structure.

Reviewer: ```suggestion
The allotropy library has a collection of data parsers, each of which should follows a general structure.

The idea of this structure is to ease the understanding of any parser in the library by defining the
same basic building blocks.

## General structure

```mermaid
classDiagram
      class ParserFactory {
          - Parser: create()
      }
      class Parser {
          - Model: parse(raw_contents: IOBase, file_name: str)
      }
      class StructureElements{
          + Any: attributes
          + StructureElement: <static> create(lines_reader: LinesReader)
      }
      class LinesReader{
          - list[str]: contents
          - str: pop()
          - str: get()
      }
      LinesReader <-- Parser
      LinesReader <-- StructureElements
      StructureElements <-- Parser
      Parser <-- ParserFactory
```

> - **LinesReader**: Default class to read raw data file.

Reviewer: ```suggestion
- **LinesReader**: Default class to read raw data file, a specific implementation may use or implement an alternative reader for its use case.
```

> - **Structure Elements**: Representation of the specific data structure of each parser.

Reviewer: ```suggestion
- **Structure Elements**: Representation of the specific data structure of the files read by each parser.
```

- **Parser**: Entry point and main controller of the parser.
- **ParserFactory**: Default constructor for any parser in the library.

## Behavior

The first step of the process is to use the **ParserFactory** to instantiate the desired **Parser**.

> This **Parser** receive a IO stream containing the raw information of the input file, in most cases

Reviewer: ```suggestion
This **Parser** receives an IO stream containing the raw information of the input file, in most cases
```

the parsers can use a **LinesReader** instance to perform the decoding and allow an easy reading
of every part of the file.

The **Parser** then invokes the **create** method of the structure elements to actually construct
the python representation of the input file (**Structure Elements**). This structure is a set
of composed **dataclass** definitions that contains all the elements needed to build the actual
output json schema. The **create** method of this **Structure Elements** usually receive an
instance of a **LinesReader** to ease the extraction of the information that they need. Finally the
 **Parser** builds the json schema using the information encoded in the **Structure Elements**.

# Calculated documents

> The intent of the calculated documents is to give information about which parts of a document are used in the calculation of another one.

Reviewer: ```suggestion
The intent of the calculated documents is to give information about which parts of a document are used in the calculation other parts.
```

# allotropy parser structure

> The allotropy library has a collection of data parsers, each of which should follows a general structure.

Reviewer: ```suggestion
The allotropy library has a collection of data parsers, each of which follows a general structure.
```

Reviewer: Missed a word here, sorry!
